[Identity] 2.0.0-beta.1 release date (#14405)
Azure/azure-sdk-for-js: sdk/identity/identity/CHANGELOG.md

# Release History

## 2.0.0-beta.1 (2021-03-23)

> This release continues with the changes from `1.2.4` and `1.2.4-beta.1`.

Reviewer: We don;t have to call out that v2 will have changes from v1.2.4. That is expected.
The call out for 1.2.4-beta.1 was added because the feature added there never went GA in any of the v1 releases.

I would recommend we simplify things and reduce confusion by removing this entire line and including the changelog from 1.2.4-beta.1 directly here for the 2.0.0-beta.1 release. The only thing there is about the auth flow and that is an important one and should get attention here

### Breaking changes

> - The `getToken` methods will now never return `null`. If a token is not available, we will return a rejected promise.

Reviewer: This generally has no affect on the end user who we expect to just pass the credential. The main person who should update their code is Azure Core or any package that calls `getToken` directly. Did we update all those cases to adjust to this change?

Also, a small suggestion on re-wording the line here
```suggestion
-  If a token is not available, the promise returned by the `getToken` method on credentials will be rejected. Previously, the promise resolved with `null`.
```

Author: I'll add that suggestion separately to avoid resolving your comment.

TypeScript didn't have issues with this API change. Nothing broke. Not sure if I should go see if anything needs cleanup if the compiler doesn't complain 😅

Reviewer: We should do more than just check that compiler doesnt complain. There might be code that checks for `null` case to determine whether there was a token or not. 

Also, in the identity.api.md file, I still see a lot of credential classes that have the getToken() method returning `Promise<AccessToken | null>`. Shouldnt that change?

Also, I recall some very old conversations around why we prefer returning null rather than throwing so that chaining becomes simpler. Is that no longer the case? Do we have language alignment here? Perhaps linking to the PR or issue that has more details will help

Author: I need to align these other PRs as well, that is true. I'll make an issue to align the `getToken` methods across the board.

Author: The | null call useful a while ago before we had the CredentialUnavailable error. Now it's not necessary and there's no logical path that results in a null token.

Author: Issue: https://github.com/Azure/azure-sdk-for-js/issues/14421

Author: I've made a PR to align the other credentials on Identity. The description of the PR states why I believe we don't need to update the rest of the codebase: https://github.com/Azure/azure-sdk-for-js/pull/14428

> - `InteractiveBrowserCredential` for the browser now requires a client ID to be provided.

Reviewer: Can we add a one liner as to why client id is so important that we made it mandatory?

Author: I'll push this:

> After we identified that the default client ID used for Node JS was not viable for the browser, we've made the client ID parameter required on the browser version of `InteractiveBrowserCredential`.

Reviewer: sounds good
Nit: Use `Node.js` instead of `Node JS` as per the way it is branded. For example, see https://nodejs.org/en/

> - `InteractiveBrowserCredential` now only has `loginStyle` and `flow` in the optional parameters when the credential is bundled for browsers. This reflects the intended behavior, since the `loginStyle` has no effect on NodeJS.

Reviewer: This is not clear to me....

> `InteractiveBrowserCredential` now only has `loginStyle` and `flow` in the optional parameters when the credential is bundled for browsers.

This tells me that there was more than loginStyle and flow before, but we don't mention what that is

> This reflects the intended behavior, since the `loginStyle` has no effect on NodeJS

This sentence would have made sense if we were saying that `loginStyle` is no longer available when using in node. But that is not what the previous sentence said

Author: `loginStyle` was available as a parameter on node but didn't do anything 🤔

Author: What about this?

> While the NodeJS version of `InteractiveBrowserCredential` allowed the `loginStyle` and `flow` optional properties to be passed through, these are only useful in the browser version of the credential. For that purpose, we've removed these parameters from the Node JS version of this credential.

Reviewer: Same for `flow`? Then this should say

> The `loginStyle` and `flow` options to the constructor for `InteractiveBrowserCredential` will now be applicable only when used in browser as this does not apply to node.js

> - Removed the `postLogoutRedirectUri` from the optional properties of the `InteractiveBrowserCredential`, to align with other languages.

Reviewer: Do the other languages not have it because it is not required at all or just that they didnt have time to get to it yet?

Author: It's not required at all.

Reviewer: Then, I would add that as the reason rather than alignment with other languages

Author: What about?

> Removed the `postLogoutRedirectUri` from the optional properties of the `InteractiveBrowserCredential`. This property is not required by our dependencies and it's not provided by the Identity SDKs in other languages.

Reviewer: Looks like we used to pass this to the old msal. Does the new `msal-browser` not take this option anymore?

Author: It takes the option, but Scott requested me to remove this from the public API since none of the other languages took this in consideration. It's not needed for our use cases (but it's not useless. It is in fact useful)

> - Added `@azure/msal-node-extensions` [1.0.0-alpha.6](https://www.npmjs.com/package/@azure/msal-node-extensions/v/1.0.0-alpha.6) as a dependency.

Reviewer: Why is this a breaking change?

Author: I'll move it to features since it's going to be an optional dependency. My concern was about the setup needed for this package. On windows, devs need a 6GB dependency to work with this.

  - This dependency requires installing `node-gyp`, and on Linux it also requires installing `libsecret`. Read more here: [link](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/msal-node-extensions#building).

> - `DefaultAzureCredential`'s implementation for browsers was simplified to throw a simple error instead of trying credentials that were never supported on the browser.

Reviewer: Why is this a breaking change? 
Also, a suggested edit:

```suggestion
- `DefaultAzureCredential`'s implementation for browsers is simplified to throw the `BrowserNotSupportedError` in its constructor. Previously, we relied on getting the same error from trying to instantiate the different  credentials that `DefaultAzureCredential` supports in node.js.
```

Author: I'll move it to features.

  - To use Identity in the browser, please use the `InteractiveBrowserCredential` directly.

### New updates and features

- Documentation was added to elaborate on how to configure an AAD application to support `InteractiveBrowserCredential`: [docuemntation link](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/interactive-browser-credential.md).
- Updated the `@azure/msal-node` dependency to `^1.0.0`.
- For the `InteractiveBrowserCredential` for node, replaced the use of the `express` module with a native http server for Node, shrinking the resulting identity module considerably.

> - `DeviceCodeCredential` now can receive its optional parameters as a single parameter object.

Reviewer: Just saw the two constructor overloads we now have:

![image](https://user-images.githubusercontent.com/16890566/112065813-929f6280-8b22-11eb-94c2-5196056bf744.png)

Given that all the parameters in the second one is already part of the `DeviceCodeCredentialOptions`, why have the second overload at all?

Author: I'll make an issue.

Author: I want to make sure I align with other languages for this method. We'll have an API alignment meeting on Wednesday.

Author: Issue: https://github.com/Azure/azure-sdk-for-js/issues/14420

Reviewer: API alignment says that all languages support creating the DeviceCodeCredential using optional parameters
- tenantId
- clientId
- userPromptCallback

JS guidelines says that all optional arguments in the signature should be combined in the options bag or you have constructor overloads for each combination

So, unless you are planning to have these overloads for the constructor as well, I don't see why need the one above
- tenantId, options
- clientId, options
- tenantId, clientId, options
- userPromptCallBack, options
- tenantId, userPromptCallback, options
- clientId, userPromptCallback, options
- tenantId, clientId, userPromptCallBack, options

Author: I'll add the change as part of this PR then.

Author: I added it as part of this PR.

- Refactored our use of MSAL to better centralize the handling of inputs, outputs and errors.
- Migrated the `InteractiveBrowserCredential`, `DeviceCodeCredential`, `ClientSecretCredential`, `ClientCertificateCredential` and `UsernamePasswordCredential` to the latest MSAL.

>   - These credentials will appear unchanged for existing use cases.
>   - However, this update improves caching of tokens, significantly reducing the number of network requests.

Reviewer: The change we made here adds no additional value. I would recommend reverting

```suggestion
  - This update improves caching of tokens, significantly reducing the number of network requests.
```

> - A new parameter named `tokenCachePersistenceOptions` can be specified on the credentials `InteractiveBrowserCredential`, `DeviceCodeCredential` and `UsernamePasswordCredential` to aneable the persitence caching.

Reviewer: ```suggestion
- A new parameter named `tokenCachePersistenceOptions` can be specified on the credentials `InteractiveBrowserCredential`, `DeviceCodeCredential` and `UsernamePasswordCredential` to enable the persistence caching.
```
